### PR TITLE
Feat/Generate nonce and vub based on notifications information

### LINKED ISSUE #844: Generate `VUB` and `nonce` based on notification transaction hash
Related to https://github.com/nspcc-dev/neofs-node/issues/843
Blocked by https://github.com/nspcc-dev/neo-go/issues/2190

With transaction hash it will be possible to calculate the same but unique `VUB` and `nonce` fields when Alphabet node create notary transaction as the reaction for notification event.

- [ ] Adopt new notification event structure
- [ ] Refactor clients / static clients / wrappers or whatever needed to pass data to calculate `VUB` and `nonce` for each notary request
- [ ] Calculate `nonce` based on transaction hash for all notary transaction produced by notifications
- [ ] Calculate `VUB` based on the block where transaction appeared, instead of block height rounding

Preliminary list of affected events:
- bind/unbind keys from NeoFS contract,
- set config from NeoFS contract,
- update event from main chain role management contract (and all governance in general, including updates based on epoch tick)

### PR DESCRIPTION
**Pull request by carpawell**
This PR refactors morph wrappers to group their call parameters. `InvokePrmOptional` is public in all level parameters to save us from too much useless copy-paste of `TxHash()` method. It will be considered to be private after moving to `neofs-sdk-go` and reducing the number of wrapper levels. 

`InvokePrmOptional` was also planned to store optional `transaction.Transaction` that is going to be signed by Alphabet. So after that PR, there will be another that deletes `NotarySignAndInvokeTX`and its usage.

Closes #844.

### CODE REVIEW

File: pkg/morph/client/static.go [hunk 1]
```diff
@@ -71,34 +72,100 @@ func (s StaticClient) Morph() *Client {
...
+	method string
+	args   []interface{}
+
+	// optional parameters
+	InvokePrmOptional
+}
+
+// InvokePrmOptional groups optional parameters of the Invoke operation.
+type InvokePrmOptional struct {
+
```
Comment (alexvanin) [inline pkg/morph/client/static.go:87]: redundant empty line

File: pkg/innerring/processors/netmap/process_epoch.go [hunk 2]
```diff
@@ -21,6 +19,8 @@ func (np *Processor) processNewEpoch(event netmapEvent.NewEpoch) {
 		np.epochState.SetEpochDuration(epochDuration)
 	}
 
+	epoch := ev.EpochNumber()
+
```
Comment (alexvanin) [inline pkg/innerring/processors/netmap/process_epoch.go:23]: Why move?

File: pkg/innerring/state.go [hunk 3]
```diff
@@ -104,9 +106,21 @@ func (s *Server) voteForSidechainValidator(validators keys.PublicKeys) error {
...
+		vub   *uint32
+		err   error
+	)
+
+	if prm.Hash != nil {
+		nonce, vub, err = s.morphClient.CalculateNonceAndVUB(*prm.Hash)
+		if err != nil {
+			return fmt.Errorf("could not calculate nonce and `validUntilBlock` values: %w", err)
+		}
+	}
```
Comment (alexvanin) [inline pkg/innerring/state.go:120]: Maybe some methods do not require `NonceAndVUB` calculations like this one. But we can keep it for similarity with others notary invocations.

File: pkg/morph/client/notary.go [hunk 4]
```diff
@@ -774,3 +827,19 @@ func CalculateNotaryDepositAmount(c *Client, gasMul, gasDiv int64) (fixedn.Fixed
 
 	return fixedn.Fixed8(depositAmount), nil
 }
+
+// CalculateNonceAndVUB calculates nonce and ValidUntilBlock values
+// based on transaction hash. Uses MurmurHash3.
+func (c *Client) CalculateNonceAndVUB(hash util.Uint256) (uint32, *uint32, error) {
```
Comment (alexvanin) [inline pkg/morph/client/notary.go:833]: There is no valid case to return `vub` as a pointer from this function.

File: pkg/morph/client/notary.go [hunk 5]
```diff
@@ -427,9 +474,15 @@ func (c *Client) notaryInvoke(committee, invokedByAlpha bool, contract util.Uint
...
 
-	until, err := c.notaryTxValidationLimit()
-	if err != nil {
-		return err
+	var until uint32
+
+	if vub != nil {
+		until = *vub
+	} else {
+		until, err = c.notaryTxValidationLimit()
```
Comment (alexvanin) [inline pkg/morph/client/notary.go:482]: I would rather keep `vub` as a `uint32` type and use `c.notaryTxValidationLimit()` at `NotaryInvokeNotAlpha()` function. Are there any other cases with nil `vub`?

**Review by alexvanin — COMMENTED** [review]
Rebase to master branch to check tests.

File: pkg/morph/client/netmap/wrapper/config.go [hunk 6]
```diff
@@ -168,9 +170,39 @@ func (w *Wrapper) readStringConfig(key string) (string, error) {
...
 // SetConfig sets config field.
-func (w *Wrapper) SetConfig(id, key []byte, value interface{}) error {
-	return w.client.SetConfig(id, key, value)
+func (w *Wrapper) SetConfig(prm SetConfigPrm) error {
+	args := netmap.SetConfigPrm{}
+
+	args.SetID(prm.id)
+	args.SetKey(prm.key)
+	args.SetValue(prm.value)
+
```
Comment (alexvanin) [inline pkg/morph/client/netmap/wrapper/config.go:204]: Don't we need `args.InvokePrmOptional = prm.InvokePrmOptional`?

File: pkg/innerring/state.go (hunk 3 above)
Comment (alexvanin) [reply pkg/innerring/state.go:121]: Yes, you are right. I thought that epoch number is enough to keep transactions unique, but Alphabet catches notifications from role management contract, so there can be multiple votes per epoch.

File: pkg/morph/client/netmap/wrapper/config.go (hunk 6 above)
Comment (carpawell) [reply pkg/morph/client/netmap/wrapper/config.go:204]: yes, we need, fixed

also, checked one more time another prm for such mistakes

Comment (alexvanin) [reply pkg/morph/client/netmap/wrapper/config.go:205]: Check pkg/morph/client/neofs/bind.go:69

File: pkg/morph/client/neofsid/wrapper/keys.go [hunk 7]
```diff
@@ -4,16 +4,28 @@ import (
 	"crypto/elliptic"
 	"fmt"
 
+	"github.com/nspcc-dev/neofs-node/pkg/morph/client"
+
```
Comment (alexvanin) [inline pkg/morph/client/neofsid/wrapper/keys.go:8]: redundant empty line

File: pkg/innerring/processors/governance/process_update.go [hunk 8]
```diff
@@ -6,15 +6,20 @@ import (
 	"sort"
 	"strings"
 
+	"github.com/nspcc-dev/neofs-node/pkg/morph/client/netmap/wrapper"
+
```
Comment (alexvanin) [inline pkg/innerring/processors/governance/process_update.go:10]: redundant empty line

File: pkg/morph/client/balance/transfer.go [hunk 9]
```diff
@@ -43,31 +47,133 @@ func (t *TransferXArgs) SetDetails(v []byte) {
...
+func (b *BurnPrm) SetTo(to []byte) {
+	b.to = to
+}
+
+// SetAmount sets amount.
+func (b *BurnPrm) SetAmount(amount int64) {
+	b.amount = amount
+}
+
+// SetID sets ID
```
Comment (alexvanin) [inline pkg/morph/client/balance/transfer.go:118]:
```suggestion
// SetID sets ID.
```

File: pkg/morph/event/container/put.go [hunk 10]
```diff
@@ -3,8 +3,9 @@ package container
 import (
 	"fmt"
 
+	"github.com/nspcc-dev/neo-go/pkg/rpc/response/result/subscriptions"
+
```
Comment (alexvanin) [inline pkg/morph/event/container/put.go:7]: redundant line

File: pkg/morph/event/container/delete.go [hunk 11]
```diff
@@ -3,8 +3,9 @@ package container
 import (
 	"fmt"
 
+	"github.com/nspcc-dev/neo-go/pkg/rpc/response/result/subscriptions"
+
```
Comment (alexvanin) [inline pkg/morph/event/container/delete.go:7]: redundant line

File: pkg/morph/event/container/estimates.go [hunk 12]
```diff
@@ -3,6 +3,8 @@ package container
 import (
 	"fmt"
 
+	"github.com/nspcc-dev/neo-go/pkg/rpc/response/result/subscriptions"
+
```
Comment (alexvanin) [inline pkg/morph/event/container/estimates.go:7]: redundant line

File: pkg/morph/client/netmap/wrapper/config.go [hunk 13]
```diff
@@ -4,6 +4,8 @@ import (
 	"fmt"
 	"strconv"
 
+	"github.com/nspcc-dev/neofs-node/pkg/morph/client"
+
```
Comment (alexvanin) [inline pkg/morph/client/netmap/wrapper/config.go:8]: redundant line

File: pkg/morph/event/balance/lock.go [hunk 14]
```diff
@@ -3,8 +3,9 @@ package balance
 import (
 	"fmt"
 
+	"github.com/nspcc-dev/neo-go/pkg/rpc/response/result/subscriptions"
+
```
Comment (alexvanin) [inline pkg/morph/event/balance/lock.go:7]: redundant line

File: pkg/innerring/processors/neofs/process_bind.go [hunk 15]
```diff
@@ -4,6 +4,8 @@ import (
 	"crypto/elliptic"
 	"fmt"
 
+	neofsid "github.com/nspcc-dev/neofs-node/pkg/morph/client/neofsid/wrapper"
+
```
Comment (alexvanin) [inline pkg/innerring/processors/neofs/process_bind.go:8]: redundant line

File: pkg/morph/event/netmap/update_peer.go [hunk 16]
```diff
@@ -4,9 +4,10 @@ import (
 	"crypto/elliptic"
 	"fmt"
 
+	"github.com/nspcc-dev/neo-go/pkg/rpc/response/result/subscriptions"
+
```
Comment (alexvanin) [inline pkg/morph/event/netmap/update_peer.go:8]: redundant line

File: pkg/innerring/processors/container/process_container.go [hunk 17]
```diff
@@ -6,10 +6,13 @@ import (
 	"errors"
 	"fmt"
 
+	"github.com/nspcc-dev/neofs-node/pkg/morph/client/container/wrapper"
+
```
Comment (alexvanin) [inline pkg/innerring/processors/container/process_container.go:10]: redundant line

**Review by alexvanin — APPROVED** [review]
LGTM, let's see if test are going to be okay.

**Pull request merged**